trin-execution/src/era_manager.rs

```diff
@@ -23,37 +47,86 @@ impl EraManager {
         let era1_files = get_shuffled_era1_files(&http_client).await?;
         Ok(Self {
             current_era: None,
+            next_era: Arc::new(Mutex::new(None)),
             http_client,
             era1_files,
         })
     }
 
-    pub async fn get_block_by_number(&mut self, block_number: u64) -> anyhow::Result<&BlockTuple> {
+    pub async fn get_block_by_number(
+        &mut self,
+        block_number: u64,
+    ) -> anyhow::Result<&BlockWithRecoveredSenders> {
         let epoch_number = Era1::epoch_number_from_block_number(block_number);
-        let era1 = match &self.current_era {
-            Some(era1) if era1.epoch_number() == epoch_number => {
+        let blocks = match &self.current_era {
+            Some(era1) if era1[0].header.number / (BLOCK_TUPLE_COUNT as u64) == epoch_number => {
                 self.current_era.as_ref().expect("current_era to be some")
             }
             _ => self
                 .current_era
-                .insert(self.download_era1_file(epoch_number).await?),
+                .insert(self.download_processed_era(epoch_number).await?),
         };
-        Ok(&era1.block_tuples[(block_number as usize) % BLOCK_TUPLE_COUNT])
+        Ok(&blocks[(block_number as usize) % BLOCK_TUPLE_COUNT])
     }
 
-    async fn download_era1_file(&self, epoch_index: u64) -> anyhow::Result<Era1> {
+    async fn download_processed_era(
+        &self,
+        epoch_index: u64,
+    ) -> anyhow::Result<Vec<BlockWithRecoveredSenders>> {
+        info!("Fetching the next era file");
+        let start = Instant::now();
+        while self.current_era.is_some() && self.next_era.lock().is_none() {
```

Comment:
I would say that this is dangerous approach for dealing with async tasks.

For example, what if task spawn at line 109 fails (for example, network issue)? It will just finish without setting `next_era`, and this one would fail after waiting for 20 minutes.